### PR TITLE
Added variations and inference documentation for Reinforcement Learning Task

### PR DESCRIPTION
**Pull request by shreyansjainn**
Addresses https://github.com/huggingface/hub-docs/issues/357.

- Added Variations (Model Based & Model Free RL)
- Added Inference Code Snippets for:
        - Interacting with the Environment using Gymnasium
        - Training, Saving, Loading and Evaluating an agent using stable-baselines3

### CODE REVIEW

**Comment by shreyansjainn** [comment]
@merveenoyan Currently I've added Model-Based and Model-Free models only in variations, should I add Value-Based & Policy Based Variants under Model-Free as well? kindly let me know.

File: tasks/src/reinforcement-learning/about.md [hunk 1]
```diff
@@ -15,7 +15,12 @@ Reinforcement learning is the science to train computers to make decisions and t
 
 ## Task Variants 
 
-You can contribute variants of this task [here](https://github.com/huggingface/hub-docs/blob/main/tasks/src/reinforcement-learning/about.md).
+### Model Based RL
+In Model based Reinforcement Learning, we try to create a model of the environment (learn the state transition probabilities and the reward function) for finding optimal actions for the agent. Some typical examples for Model Based RL algorithms are Dynamic Programming, Value Iteration & Policy Iteration
+
+
+### Model Free RL
+In Model Free RL, agent decides on optimal actions, solely based on its experience in the environment and the reward it collects during that journey. This is one of the most commonly used algorithms specially beneficial in complex environments where modelling of state transition probabilities and reward functions are quite difficult. Some of the examples will be SARSA, Q-Learning, Actor-Critic, Proximal Policy Optimization (PPO) algorithms.
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:23]:
```suggestion
In model free reinforcement learning, agent decides on optimal actions based on its experience in the environment and the reward it collects from it. This is one of the most commonly used algorithms beneficial in complex environments, where modeling of state transition probabilities and reward functions are difficult. Some of the examples of model free reinforcement learning are SARSA, Q-Learning, actor-critic and proximal policy optimization (PPO) algorithms.

```

File: tasks/src/reinforcement-learning/about.md [hunk 2]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
 
 ## Inference
 
-You can add a small snippet [here](https://github.com/huggingface/hub-docs/blob/main/tasks/src/reinforcement-learning/about.md) that shows how to infer with `reinforcement-learning` models.
+Inference in Reinforcement learning, differs a bit from traditional supervised learning where you just deal with model and test data. In RL, once you have trained an agent in an env, you try to run the trained agent for specific no of steps to get the average reward its able to get.
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:60]:
```suggestion
Inference in reinforcement learning differs from other modalities, in which there's a model and test data. In reinforcement learning, once you have trained an agent in an environment, you try to run the trained agent for additional steps to get the average reward. 

```
I feel like you're mentioning about evaluation of agents here and not directly inference, am I correct?

File: tasks/src/reinforcement-learning/about.md [hunk 3]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
 
 ## Inference
 
-You can add a small snippet [here](https://github.com/huggingface/hub-docs/blob/main/tasks/src/reinforcement-learning/about.md) that shows how to infer with `reinforcement-learning` models.
+Inference in Reinforcement learning, differs a bit from traditional supervised learning where you just deal with model and test data. In RL, once you have trained an agent in an env, you try to run the trained agent for specific no of steps to get the average reward its able to get. 
+
+A typical training cycle comprises of gathering experience from env, training the model and running the model on test env to get average reward. Below we have showcased, how you can interact with the env using gymnasium library, train a model using stable-baselines3 and run it on test environment.
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:62]:
```suggestion
A typical training cycle consists of gathering experience from the environment, training the model, and running the model on a test environment to obtain average reward. Below there's a snippet on how you can interact with the environment using the `gymnasium` library, train a model using `stable-baselines3` and run the agent on test environment.

```

File: tasks/src/reinforcement-learning/about.md [hunk 4]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
...
+```python
+# Here we are running 20 episodes of CartPole-v1 environment, taking random actions
+import gymnasium as gym
+
+env = gym.make("CartPole-v1")
+observation, info = env.reset()
+
+for _ in range(20):
+	action = env.action_space.sample() # samples random action from action sample space
+	print("Action Take:",action)
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:73]:
```suggestion
```
trying to keep the snippet simpler

File: tasks/src/reinforcement-learning/about.md [hunk 5]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
...
+
+	observation, reward, terminated, truncated, info = env.step(action) # takes the action in the env
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:75]:
```suggestion
        # the agent takes the action 
	observation, reward, terminated, truncated, info = env.step(action)

```

File: tasks/src/reinforcement-learning/about.md [hunk 6]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
...
+
+if terminated or truncated: # if the agent reaches terminal state, we reset the environment
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:77]:
```suggestion
# if the agent reaches terminal state, we reset the environment
if terminated or truncated: 

```

File: tasks/src/reinforcement-learning/about.md [hunk 7]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
...
+	observation = env.reset()
+
+env.close()
+```
+
+```python
+# Training a PPO model on LunarLander-v2 environment using stable-baselines3 library and saving the model
+from stable_baselines3 import PPO
+
+# Environment Initialization
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:88]:
```suggestion
# initialize the environment

```

File: tasks/src/reinforcement-learning/about.md [hunk 8]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
...
+env = gym.make("LunarLander-v2")
+
+# Model Initialization
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:91]:
```suggestion
# initialize the model

```

File: tasks/src/reinforcement-learning/about.md [hunk 9]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
...
+model = PPO(policy = "MlpPolicy",
+			env = env,
+			n_steps = 1024,
+			batch_size = 64,
+			n_epochs = 4,
+			verbose = 1)
+
+# trains the model for 1000 time steps
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:99]:
```suggestion
# train the model for 1000 time steps

```

File: tasks/src/reinforcement-learning/about.md [hunk 10]
```diff
@@ -52,7 +57,77 @@ Observations and states are the information our agent gets from the environment.
...
+model.learn(total_timesteps = 1000)
+
+# Saving the model in desired directory
+model_name = "PPO-LunarLander-v2"
+model.save(model_name)
+```
+
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:106]: Maybe you can add
```suggestion
Below code shows how to make inference with an agent trained using `stable-baselines3`
```

**Review by merveenoyan — COMMENTED** [review]
Great PR! @simoninithomas can you also check it out?

File: tasks/src/reinforcement-learning/about.md (hunk 2 above)
Comment (shreyansjainn) [reply tasks/src/reinforcement-learning/about.md:62]: yes, i'll add a code snippet for inference as well. missed out on that

**Review by simoninithomas — APPROVED** [review]
LGTM Thanks for adding this part 🤗

File: tasks/src/reinforcement-learning/about.md [hunk 11]
```diff
@@ -15,7 +15,14 @@ Reinforcement learning is the science to train computers to make decisions and t
 
 ## Task Variants 
 
-You can contribute variants of this task [here](https://github.com/huggingface/hub-docs/blob/main/tasks/src/reinforcement-learning/about.md).
+### Model Based RL
+In model based reinforcement learning techniques intend to create a model of the environment, learn the state transition probabilities and the reward function, to find the optimal action. Some typical examples for model based reinforcement learning algorithms are dynamic programming, value iteration and policy iteration.
+
+
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:21]:
```suggestion
```

File: tasks/src/reinforcement-learning/about.md [hunk 12]
```diff
@@ -52,7 +59,107 @@ Observations and states are the information our agent gets from the environment.
 
 ## Inference
 
-You can add a small snippet [here](https://github.com/huggingface/hub-docs/blob/main/tasks/src/reinforcement-learning/about.md) that shows how to infer with `reinforcement-learning` models.
+Inference in reinforcement learning differs from other modalities, in which there's a model and test data. In reinforcement learning, once you have trained an agent in an environment, you try to run the trained agent for additional steps to get the average reward. 
+
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:63]:
```suggestion

```

File: tasks/src/reinforcement-learning/about.md [hunk 13]
```diff
@@ -52,7 +59,107 @@ Observations and states are the information our agent gets from the environment.
...
+# if the agent reaches terminal state, we reset the environment
+if terminated or truncated: 
+
+	print("Environment is reset")
+	observation = env.reset()
+
+env.close()
+```
+
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:90]:
```suggestion

Below snippet shows how to train a PPO model on LunarLander-v2 environment using `stable-baselines3` library and saving the model
```

File: tasks/src/reinforcement-learning/about.md [hunk 14]
```diff
@@ -52,7 +59,107 @@ Observations and states are the information our agent gets from the environment.
...
+```python
+# Training a PPO model on LunarLander-v2 environment using stable-baselines3 library and saving the model
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:92]:
```suggestion

```

File: tasks/src/reinforcement-learning/about.md [hunk 15]
```diff
@@ -52,7 +59,107 @@ Observations and states are the information our agent gets from the environment.
...
+
+model = PPO(policy = "MlpPolicy",
+			env = env,
+			n_steps = 1024,
+			batch_size = 64,
+			n_epochs = 4,
+			verbose = 1)
+
+# train the model for 1000 time steps
+
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:109]:
```suggestion
```

File: tasks/src/reinforcement-learning/about.md [hunk 16]
```diff
@@ -52,7 +59,107 @@ Observations and states are the information our agent gets from the environment.
...
+env = gym.make("LunarLander-v2")
+# Loading the saved model
+model = PPO.load("PPO-LunarLander-v2",env=env)
+
+# Initializating the evaluation environment
+eval_env = gym.make("LunarLander-v2")
+
+# Running the trained agent on eval_env for 10 time steps and getting the mean reward
+mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes = 10,
+										  deterministic=True)
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:132]:
```suggestion
										  deterministic=True)
```

File: tasks/src/reinforcement-learning/about.md [hunk 17]
```diff
@@ -52,7 +59,107 @@ Observations and states are the information our agent gets from the environment.
...
+env = model.get_env()
+
+obs = env.reset()
+for i in range(1000):
+	# getting action predictions from the trained agent
+	action, _states = model.predict(obs, deterministic=True)
+	
+	# taking the predicted action in the environment to observe next state and rewards
+    obs, rewards, dones, info = env.step(action)
+
```
Comment (merveenoyan) [inline tasks/src/reinforcement-learning/about.md:156]:
```suggestion
```

**Review by merveenoyan — APPROVED** [review]
Looks good, let's get it merged!!!

**Pull request merged**